File: src/main/java/net/nerol/bot_simulator/Environment.java
package net.nerol.bot_simulator;

import net.nerol.bot_simulator.minecraft.world.entity.PvPBot;
import net.nerol.bot_simulator.minecraft.world.entity.ai.attributes.AttributeType;
import net.nerol.bot_simulator.minecraft.world.item.ItemStack;
import net.nerol.bot_simulator.minecraft.world.item.ItemType;
import net.nerol.bot_simulator.minecraft.world.item.enchantments.Enchantment;
import net.nerol.bot_simulator.minecraft.world.item.enchantments.EnchantmentType;
import net.nerol.bot_simulator.minecraft.world.phys.PhysicsEngine;
import net.nerol.bot_simulator.minecraft.world.phys.RayTrace;

import java.util.Random;

public class Environment {
    // MC's MOVEMENT_SPEED attribute (default 0.1) corresponds to ~0.21585 blocks/tick
    // walking equilibrium under 0.91 horizontal friction. Per-tick impulse therefore
    // equals MOVEMENT_SPEED * (0.21585 / 0.1) * (1 - 0.91) = MOVEMENT_SPEED * 0.1943.
    // Sprint multiplies the walking velocity by 1.3.
    private static final double WALK_IMPULSE_PER_SPEED = 0.1943;
    private static final double SPRINT_MULTIPLIER = 1.3;

    // MC's JUMP_STRENGTH default (~0.42 blocks/tick initial vertical velocity).
    private static final double JUMP_VELOCITY = 0.42;

    public PvPBot bot1;
    public PvPBot bot2;
    private PhysicsEngine physics = new PhysicsEngine();
    public Environment() {
        bot1 = new PvPBot();
        bot2 = new PvPBot();

        reset();
    }

    public void executeAction(Action action) {
        /*System.out.printf(
                "Before: player=(%.2f, %.2f) enemy=(%.2f, %.2f)%n",
                bot1.Pos.x, bot1.Pos.z, bot2.Pos.x, bot2.Pos.z
        );*/

        // Clear both per-tick hit flags; they get set true again only if a hit lands this tick.
        bot1.wasHit = false;
        bot2.wasHit = false;

        switch (action) {
            case SPRINT:
                if (!bot1.walking_back) {
                    bot1.sprinting = !bot1.sprinting;
                }
                break;
            case MOVE_FORWARD:
                if (!bot1.sprinting) {
                    bot1.walking_forward = !bot1.walking_forward;
                }
                break;
            case MOVE_BACK:
                if (!bot1.walking_forward && !bot1.sprinting) {
                    bot1.walking_back = !bot1.walking_back;
                }
                break;

            case STRAFE_LEFT:
                if (!bot1.strafing_right) {
                    bot1.strafing_left = !bot1.strafing_left;
                }
                break;

            case STRAFE_RIGHT:
                if (!bot1.strafing_left) {
                    bot1.strafing_right = !bot1.strafing_right;
                }
                break;

            case ATTACK:
                attack();
                break;

            case TURN_LEFT_45:
                turn(-45);
                break;

            case TURN_RIGHT_45:
                turn(45);
                break;

            case TURN_LEFT_90:
                turn(-90);
                break;

            case TURN_RIGHT_90:
                turn(90);
                break;

            case JUMP:
                jump();
                break;
        }

        if (bot1.sprinting) setSprinting();
        if (bot1.walking_forward) setWalking();
        if (bot1.walking_back) setBackward();
        if (bot1.strafing_left) setStrafeLeft();
        if (bot1.strafing_right) setStrafeRight();

        updatebot2();
        physics.update(bot1);
        physics.update(bot2);

        /*System.out.printf(
                "After: player=(%.2f, %.2f) enemy=(%.2f, %.2f)%n",
                bot1.Pos.x, bot1.Pos.z, bot2.Pos.x, bot2.Pos.z
        );*/
    }

    // --- Movement impulses driven by the MOVEMENT_SPEED attribute ---
    void setSprinting() {
        double impulse = bot1.attributes.get(AttributeType.MOVEMENT_SPEED) * WALK_IMPULSE_PER_SPEED * SPRINT_MULTIPLIER;
        bot1.Motion.x += Math.cos(Math.toRadians(bot1.getYaw())) * impulse;
        bot1.Motion.z += Math.sin(Math.toRadians(bot1.getYaw())) * impulse;
    }

    void setWalking() {
        double impulse = bot1.attributes.get(AttributeType.MOVEMENT_SPEED) * WALK_IMPULSE_PER_SPEED;
        bot1.Motion.x += Math.cos(Math.toRadians(bot1.getYaw())) * impulse;
        bot1.Motion.z += Math.sin(Math.toRadians(bot1.getYaw())) * impulse;
    }
    void setBackward() {
        double impulse = -bot1.attributes.get(AttributeType.MOVEMENT_SPEED) * WALK_IMPULSE_PER_SPEED;
        bot1.Motion.x += Math.cos(Math.toRadians(bot1.getYaw())) * impulse;
        bot1.Motion.z += Math.sin(Math.toRadians(bot1.getYaw())) * impulse;
    }
    void setStrafeLeft() {
        double impulse = bot1.attributes.get(AttributeType.MOVEMENT_SPEED) * WALK_IMPULSE_PER_SPEED;
        bot1.Motion.x += Math.cos(Math.toRadians(bot1.getYaw() - 90)) * impulse;
        bot1.Motion.z += Math.sin(Math.toRadians(bot1.getYaw() - 90)) * impulse;
    }
    void setStrafeRight() {
        double impulse = bot1.attributes.get(AttributeType.MOVEMENT_SPEED) * WALK_IMPULSE_PER_SPEED;
        bot1.Motion.x += Math.cos(Math.toRadians(bot1.getYaw() + 90)) * impulse;
        bot1.Motion.z += Math.sin(Math.toRadians(bot1.getYaw() + 90)) * impulse;
    }

    void attack() {
        performAttack(bot1, bot2);
    }

    /** Faithful port of the mod's {@code PvPBot.attack(Entity)}: attack-strength scaling,
     *  separate physical vs. enchantment (Sharpness) damage, full-strength gating, critical
     *  hits, and the sprint knockback bonus. The deployed bot (ActionPack.leftClick) only
     *  swings at strength > 0.9, and an out-of-reach swing hits air without consuming the
     *  cooldown — both mirrored here. */
    void performAttack(PvPBot attacker, PvPBot defender) {
        // attackStrengthScale = getAttackStrengthScale(0.5F): cooldown charge in [0,1].
        double scale = Math.max(0.0, Math.min(1.0, attacker.attackCharge));

        boolean fullStrengthAttack = scale > 0.9;
        if (!fullStrengthAttack) {
            return; // leftClick drops a non-full swing before it even swings.
        }

        // Entity-pick raycast (mirrors ActionPack.leftClick): the swing only lands if a ray
        // from the attacker's eye along its look direction clips the defender's hitbox within
        // reach. A swing at air — out of range OR not aimed at the target — hits nothing and
        // (as before) does not consume the cooldown.
        if (!RayTrace.canHit(attacker, defender)) {
            return;
        }

        // baseDamage = weapon's physical ATTACK_DAMAGE; magicBoost = the Sharpness portion,
        // scaled by attack strength (getEnchantedDamage - baseDamage).
        double baseDamage = attacker.equipment.SelectedItem.type.attackDamage;
        double enchantedDamage = attacker.equipment.SelectedItem.getAttackDamage();
        double magicBoost = scale * (enchantedDamage - baseDamage);

        // Cooldown ramp on the physical portion: baseDamageScaleFactor() = 0.2 + 0.8*scale^2.
        baseDamage *= 0.2 + 0.8 * scale * scale;

        boolean knockbackAttack = attacker.sprinting && fullStrengthAttack;

        // Critical hit multiplies ONLY the physical portion; magicBoost is added afterwards.
        boolean criticalAttack = fullStrengthAttack && canCriticalAttack(attacker);
        if (criticalAttack) {
            baseDamage *= 1.5;
        }

        double totalDamage = baseDamage + magicBoost;

        // hurtOrSimulate analogue: reduce by the defender's Protection multiplier.
        double damageTaken = totalDamage * defender.equipment.getProtectionDamageMultiplier();
        if (damageTaken > 0.0) {
            defender.Health -= (float) damageTaken;

            // causeExtraKnockback amount = Knockback-enchant + (sprint ? 0.5 : 0). The base
            // 0.4 hit reaction is applied inside the knockback system.
            double enchantKnockback =
                    attacker.equipment.SelectedItem.getEnchantmentLevel(EnchantmentType.KNOCKBACK) * 0.5;
            double extraKnockback = enchantKnockback + (knockbackAttack ? 0.5 : 0.0);
            physics.getKnockbackSystem().applyHitKnockback(attacker, defender, extraKnockback);
        }

        // resetAttackStrengthTicker: the charge is consumed only on a real (in-reach) attack.
        attacker.attackCharge = 0.0;
    }

    /** Subset of the mod's {@code canCriticalAttack} the simulator can observe: the attacker
     *  must be descending (fall_distance > 0), airborne, and not sprinting. The remaining
     *  guards (not on a ladder / in water / a passenger; target is a LivingEntity) always
     *  hold in this 1v1 setup. */
    boolean canCriticalAttack(PvPBot attacker) {
        return attacker.fall_distance > 0.0f
                && !attacker.onGround
                && !attacker.sprinting;
    }
    void jump() {
        if (bot1.onGround) {
            bot1.Motion.y = JUMP_VELOCITY;
            bot1.onGround = false;
        }
    }
    void turn(int yawDelta) {
        bot1.Rotation.x += yawDelta;
        bot1.Rotation.x = (bot1.Rotation.x % 360 + 360) % 360;
    }


    void lookAt(PvPBot self, PvPBot target) {
        double dx = target.Pos.x - self.Pos.x;
        double dy = target.Pos.y - self.Pos.y;
        double dz = target.Pos.z - self.Pos.z;

        self.Rotation.x = (float) Math.toDegrees(Math.atan2(dz, dx));
        self.Rotation.y = (float) -Math.toDegrees(Math.atan2(dy, Math.sqrt(dx * dx + dz * dz)));
    }

    void stopMovement() {
        bot1.Motion.x = 0;
        bot1.Motion.y = 0;
    }

    // --- bot2 s-tap combo state ---
    // Ticks remaining in the post-hit backward tap. The s-tap is a brief retreat between
    // hits that keeps bot2 from overshooting bot1 while its attack recharges, then lets it
    // sprint back in so every re-approach lands a fresh sprint-knockback hit.
    private int bot2StapTicks = 0;

    private static final double S_TAP_MEAN = 3;
    private static final double S_TAP_STDDEV = 0.75;
    private static final int S_TAP_MAX = 5;
    private static final long S_TAP_SEED = 42L;
    private final Random random = new Random(S_TAP_SEED);

    void updatebot2() {
        // S-tapping combo trainer: bot2 sprint-chases bot1, lands a sprint-knockback hit,
        // then taps backward for a couple ticks (the s-tap) before sprinting back in —
        // keeping bot1 pinned in a knockback loop. Damage, criticals and the +0.5 sprint-
        // knockback bonus all run through performAttack, exactly as bot1's attacks do.
        // bot2's movement is applied inline here so this stays self-contained to bot2.

        // Face bot1 every tick. Yaw drives both movement and the attack ray; pitch keeps the
        // ray aimed at bot1's hitbox when there's a vertical gap (e.g. bot1 jumps for a crit),
        // so the 3D raytrace still lands. Movement impulses below use yaw only, so pitch is
        // purely an aiming term.
        double dx = bot1.Pos.x - bot2.Pos.x;
        double dy = bot1.Pos.y - bot2.Pos.y;
        double dz = bot1.Pos.z - bot2.Pos.z;
        bot2.Rotation.x = (float) Math.toDegrees(Math.atan2(dz, dx));
        bot2.Rotation.y = (float) -Math.toDegrees(Math.atan2(dy, Math.sqrt(dx * dx + dz * dz)));
        double yaw = Math.toRadians(bot2.getYaw());

        double walkImpulse = bot2.attributes.get(AttributeType.MOVEMENT_SPEED) * WALK_IMPULSE_PER_SPEED;
        if (bot2StapTicks > 0) {
            // Post-hit s-tap: walk backward briefly. The landed hit's knockback already
            // cleared bot2's sprint, so this just spaces the approach and re-times the next
            // sprint-in to arrive in reach right as the attack finishes recharging.
            bot2StapTicks--;
            bot2.sprinting = false;
            bot2.walking_back = true;
            bot2.Motion.x -= Math.cos(yaw) * walkImpulse;
            bot2.Motion.z -= Math.sin(yaw) * walkImpulse;
        } else {
            // Chase: keep sprinting straight at bot1 until the raytrace says we can connect.
            // sprinting must be true at swing time for performAttack to add the +0.5 sprint-
            // knockback bonus, so committing to the sprint-in is what locks in the combo.
            bot2.sprinting = true;
            bot2.walking_back = false;
            double sprintImpulse = walkImpulse * SPRINT_MULTIPLIER;
            bot2.Motion.x += Math.cos(yaw) * sprintImpulse;
            bot2.Motion.z += Math.sin(yaw) * sprintImpulse;
        }

        // Only swing once we can actually land it (in reach AND aimed) — the combo lock-in.
        // Until then bot2 just keeps sprinting in above; no wasted swings at air. The first
        // connecting sprint-hit consumes the charge (a drop signals the hit landed), which
        // kicks off the s-tap and the knockback loop that sustains the combo.
        if (RayTrace.canHit(bot2, bot1)) {
            double chargeBeforeSwing = bot2.attackCharge;
            performAttack(bot2, bot1);
            if (bot2.attackCharge < chargeBeforeSwing) {
                // Jittered s-tap length: round a Gaussian draw and clamp to [0, S_TAP_MAX].
                double sampled = S_TAP_MEAN + random.nextGaussian() * S_TAP_STDDEV;
                bot2StapTicks = (int) Math.round(Math.max(1.0, Math.min(S_TAP_MAX, sampled)));
            }
        }
    }

    /** Episode is over when either bot has been reduced to 0 HP. */
    public boolean isEpisodeOver() {
        return bot1.Health <= 0 || bot2.Health <= 0;
    }

    // --- Observations (you implement these) ---
    public State getCurrentState() {
        int distance = computeDistanceBucket();
        int direction = computeDirectionBucket();
        return new State(distance, direction);
    }

    int computeDistanceBucket() {
        // Full 3D distance from bot1 to bot2, not from world origin. Includes the vertical
        // gap so a jump/crit separation registers, matching the 3D entity-pick raytrace.
        // Same convention the MC LiveController will use, so the Q-table transfers cleanly.
        double dx = bot2.Pos.x - bot1.Pos.x;
        double dy = bot2.Pos.y - bot1.Pos.y;
        double dz = bot2.Pos.z - bot1.Pos.z;
        double dist = Math.sqrt(dx * dx + dy * dy + dz * dz);

        if (dist < 1.66666) return 0;
        if (dist <= 3) return 1;
        return 2;
    }
    int computeDirectionBucket() {
        // Target's bearing relative to bot1's facing, bucketed into 8 sectors of 45 degrees.
        // 0 = target dead ahead, 2 = right, 4 = behind, 6 = left. The +22.5 offset centers
        // bucket 0 on "directly in front" so a small mis-aim still reads as FRONT.
        double dx = bot2.Pos.x - bot1.Pos.x;
        double dz = bot2.Pos.z - bot1.Pos.z;
        double bearingToTarget = Math.toDegrees(Math.atan2(dz, dx));
        double relative = ((bearingToTarget - bot1.getYaw()) % 360 + 360 + 22.5) % 360;
        return (int)(relative / 45.0); // 0..7
    }
    double distanceSquaredTobot2() {
        double dx = bot2.Pos.x - bot1.Pos.x;
        double dy = bot2.Pos.y - bot1.Pos.y;
        double dz = bot2.Pos.z - bot1.Pos.z;

        return dx * dx + dy * dy + dz * dz;
    }
    public void reset() {
        // player
        bot1.Pos.x = 0;
        bot1.Pos.y = 0;
        bot1.Pos.z = 0;
        bot1.Motion.x = 0;
        bot1.Motion.y = 0;
        bot1.Motion.z = 0;
        bot1.Rotation.x = 0.0f;
        bot1.onGround = true;
        bot1.attackCharge = 1.0;
        bot1.sweepingAttack = false;
        bot1.wasHit = false;
        bot1.sprinting = false;
        bot1.walking_forward = false;
        bot1.strafing_left = false;
        bot1.strafing_right = false;
        bot1.walking_back = false;

        bot1.Health = 20.0f;
        bot1.equipment.SelectedItem = new ItemStack(ItemType.DIAMOND_SWORD);
        bot1.equipment.head  = new ItemStack(ItemType.DIAMOND_HELMET,     new Enchantment(EnchantmentType.PROTECTION, 4));
        bot1.equipment.chest = new ItemStack(ItemType.DIAMOND_CHESTPLATE, new Enchantment(EnchantmentType.PROTECTION, 4));
        bot1.equipment.leg   = new ItemStack(ItemType.DIAMOND_LEGGINGS,   new Enchantment(EnchantmentType.PROTECTION, 4));
        bot1.equipment.feet  = new ItemStack(ItemType.DIAMOND_BOOTS,      new Enchantment(EnchantmentType.PROTECTION, 4));
        resetAttributesFromEquipment(bot1);

        // bot2
        bot2.Pos.x = 10;
        bot2.Pos.y = 0;
        bot2.Pos.z = 0;
        bot2.Motion.x = 0;
        bot2.Motion.y = 0;
        bot2.Motion.z = 0;
        bot2.Rotation.x = 180;
        bot2.onGround = true;
        bot2.attackCharge = 1.0;
        bot2.sweepingAttack = false;
        bot2.wasHit = false;
        bot2.sprinting = false;
        bot2.walking_forward = false;
        bot2.strafing_left = false;
        bot2.strafing_right = false;
        bot2.walking_back = false;

        bot2.Health = 20.0f;
        bot2.equipment.SelectedItem = new ItemStack(ItemType.DIAMOND_SWORD);
        bot2.equipment.head  = new ItemStack(ItemType.DIAMOND_HELMET,     new Enchantment(EnchantmentType.PROTECTION, 4));
        bot2.equipment.chest = new ItemStack(ItemType.DIAMOND_CHESTPLATE, new Enchantment(EnchantmentType.PROTECTION, 4));
        bot2.equipment.leg   = new ItemStack(ItemType.DIAMOND_LEGGINGS,   new Enchantment(EnchantmentType.PROTECTION, 4));
        bot2.equipment.feet  = new ItemStack(ItemType.DIAMOND_BOOTS,      new Enchantment(EnchantmentType.PROTECTION, 4));
        resetAttributesFromEquipment(bot2);
    }

    /** Repopulate attributes from the bot's current gear. Movement speed stays at the
     *  player base value; damage/speed/armor/toughness come from the equipped items. */
    private void resetAttributesFromEquipment(PvPBot bot) {
        bot.attributes.set(AttributeType.MOVEMENT_SPEED, 0.1);
        bot.attributes.set(AttributeType.ATTACK_DAMAGE, bot.equipment.SelectedItem.getAttackDamage());
        bot.attributes.set(AttributeType.ATTACK_SPEED, bot.equipment.SelectedItem.getAttackSpeed());
        bot.attributes.set(AttributeType.ARMOR, bot.equipment.getArmorPoints());
        bot.attributes.set(AttributeType.ARMOR_TOUGHNESS, bot.equipment.getArmorToughness());
    }
}

File: src/PythonTrainer/environment.py
"""
Duel environment for the neural-net trainer — a Python port of the Java
`RL/Simulator/bot_simulator/Environment.java` (+ PhysicsEngine / Knockback / RayTrace
and the bot2 s-tap AI), extended to a full 360-degree yaw + pitch action space.

bot1 is the learning agent; bot2 is the scripted s-tap opponent. Conventions match the
Java sim: xz forward = (cos yaw, sin yaw); attacks are gated by an eye->look raytrace
against the target's AABB, so the agent must actually aim (now in 3D) to land a hit.

The env is Gym-like: reset() -> obs, step(action) -> (obs, reward, done, info).
"""

import math
import numpy as np

# --- MC-like constants (mirror Environment.java) ---
MOVEMENT_SPEED = 0.1
WALK_IMPULSE_PER_SPEED = 0.1943
SPRINT_MULTIPLIER = 1.3
WALK_IMPULSE = MOVEMENT_SPEED * WALK_IMPULSE_PER_SPEED   # per-tick forward impulse

H_FRICTION = 0.91          # horizontal ground friction
GRAVITY = 0.08
V_DRAG = 0.98
JUMP_VELOCITY = 0.42

REACH = 3.0                # ENTITY_INTERACTION_RANGE
PLAYER_WIDTH = 0.6
PLAYER_HEIGHT = 1.8
EYE_HEIGHT = 1.62

ATTACK_DAMAGE = 7.0                 # diamond sword physical damage
CHARGE_PER_TICK = 1.0 / 12.5        # diamond sword attack speed (~1.6/s -> full in 12.5 ticks)
FULL_STRENGTH = 0.9
# Net damage multiplier of full diamond armor + Protection IV (approximate; the Java sim
# computes this from the Equipment, here it's a constant so a clean hit deals ~2 HP).
PROTECTION_MULT = 0.30
BASE_KNOCKBACK = 0.4
SPRINT_KNOCKBACK = 0.5

# bot2 s-tap timing (mirror Environment.java)
S_TAP_MEAN = 3.0
S_TAP_STD = 0.75
S_TAP_MAX = 5

MAX_HEALTH = 20.0

# --- Action space (discrete; small turn deltas compose to any yaw/pitch) ---
(IDLE, FORWARD, SPRINT_FORWARD, BACK, STRAFE_LEFT, STRAFE_RIGHT,
 ATTACK, JUMP, YAW_L, YAW_R, YAW_L_FINE, YAW_R_FINE, PITCH_UP, PITCH_DOWN) = range(14)
NUM_ACTIONS = 14
YAW_STEP = 15.0
YAW_STEP_FINE = 4.0
PITCH_STEP = 8.0

OBS_DIM = 13


def _wrap_deg(a):
    return (a + 180.0) % 360.0 - 180.0


def _look_vec(yaw_deg, pitch_deg):
    """Unit look vector; xz forward = (cos yaw, sin yaw), matching the Java RayTrace."""
    yaw = math.radians(yaw_deg)
    pitch = math.radians(pitch_deg)
    cp = math.cos(pitch)
    return cp * math.cos(yaw), -math.sin(pitch), cp * math.sin(yaw)


def _ray_hits(att, dfn, reach=REACH):
    """Slab-method port of RayTrace.canHit: does att's eye->look ray clip dfn's AABB?"""
    ex, ey, ez = att.x, att.y + EYE_HEIGHT, att.z
    lx, ly, lz = _look_vec(att.yaw, att.pitch)
    dx, dy, dz = lx * reach, ly * reach, lz * reach
    half = PLAYER_WIDTH / 2.0
    box = ((dfn.x - half, dfn.x + half),
           (dfn.y, dfn.y + PLAYER_HEIGHT),
           (dfn.z - half, dfn.z + half))
    origin = (ex, ey, ez)
    delta = (dx, dy, dz)
    tmin, tmax = 0.0, 1.0
    for o, d, (lo, hi) in zip(origin, delta, box):
        if abs(d) < 1e-8:
            if o < lo or o > hi:
                return False
        else:
            t1 = (lo - o) / d
            t2 = (hi - o) / d
            if t1 > t2:
                t1, t2 = t2, t1
            tmin = max(tmin, t1)
            tmax = min(tmax, t2)
            if tmin > tmax:
                return False
    return True


class Bot:
    __slots__ = ("x", "y", "z", "vx", "vy", "vz", "yaw", "pitch",
                 "health", "charge", "on_ground", "sprinting", "was_hit")

    def reset(self, x, z, yaw):
        self.x, self.y, self.z = x, 0.0, z
        self.vx = self.vy = self.vz = 0.0
        self.yaw, self.pitch = yaw, 0.0
        self.health = MAX_HEALTH
        self.charge = 1.0
        self.on_ground = True
        self.sprinting = False
        self.was_hit = False


class DuelEnv:
    def __init__(self, max_steps=1200, seed=0):
        self.max_steps = max_steps
        self.rng = np.random.default_rng(seed)
        self.bot1 = Bot()
        self.bot2 = Bot()
        self.reset()

    # ------------------------------------------------------------------ core
    def reset(self):
        self.bot1.reset(0.0, 0.0, 0.0)
        self.bot2.reset(10.0, 0.0, 180.0)
        self.steps = 0
        self._stap_ticks = 0
        self._prev_dist, self._prev_aim = self._dist(), self._aim_error()
        return self._observe()

    def step(self, action):
        self.bot1.was_hit = False
        self.bot2.was_hit = False
        hp2_before = self.bot2.health
        hp1_before = self.bot1.health

        could_hit = self.bot1.charge > FULL_STRENGTH and _ray_hits(self.bot1, self.bot2)
        self._apply_action(self.bot1, action)
        self._opponent_ai()

        self._physics(self.bot1)
        self._physics(self.bot2)
        self._recharge(self.bot1)
        self._recharge(self.bot2)

        self.steps += 1
        dmg_dealt = max(0.0, hp2_before - self.bot2.health)
        dmg_taken = max(0.0, hp1_before - self.bot1.health)
        reward = self._reward(action, dmg_dealt, dmg_taken, could_hit)
        done = self.bot1.health <= 0 or self.bot2.health <= 0 or self.steps >= self.max_steps
        return self._observe(), reward, done, {"dmg_dealt": dmg_dealt, "dmg_taken": dmg_taken}

    # ------------------------------------------------------------- actions
    def _apply_action(self, b, a):
        yaw = math.radians(b.yaw)
        if a == FORWARD:
            b.vx += math.cos(yaw) * WALK_IMPULSE
            b.vz += math.sin(yaw) * WALK_IMPULSE
        elif a == SPRINT_FORWARD:
            b.sprinting = True
            b.vx += math.cos(yaw) * WALK_IMPULSE * SPRINT_MULTIPLIER
            b.vz += math.sin(yaw) * WALK_IMPULSE * SPRINT_MULTIPLIER
        elif a == BACK:
            b.sprinting = False
            b.vx -= math.cos(yaw) * WALK_IMPULSE
            b.vz -= math.sin(yaw) * WALK_IMPULSE
        elif a == STRAFE_LEFT:
            b.sprinting = False
            b.vx += math.cos(yaw - math.pi / 2) * WALK_IMPULSE
            b.vz += math.sin(yaw - math.pi / 2) * WALK_IMPULSE
        elif a == STRAFE_RIGHT:
            b.sprinting = False
            b.vx += math.cos(yaw + math.pi / 2) * WALK_IMPULSE
            b.vz += math.sin(yaw + math.pi / 2) * WALK_IMPULSE
        elif a == ATTACK:
            self._perform_attack(self.bot1, self.bot2)
        elif a == JUMP:
            if b.on_ground:
                b.vy = JUMP_VELOCITY
                b.on_ground = False
        elif a == YAW_L:
            b.yaw = _wrap_deg(b.yaw - YAW_STEP)
        elif a == YAW_R:
            b.yaw = _wrap_deg(b.yaw + YAW_STEP)
        elif a == YAW_L_FINE:
            b.yaw = _wrap_deg(b.yaw - YAW_STEP_FINE)
        elif a == YAW_R_FINE:
            b.yaw = _wrap_deg(b.yaw + YAW_STEP_FINE)
        elif a == PITCH_UP:
            b.pitch = max(-90.0, b.pitch - PITCH_STEP)
        elif a == PITCH_DOWN:
            b.pitch = min(90.0, b.pitch + PITCH_STEP)
        # IDLE: nothing — friction/gravity still apply in _physics

    # ----------------------------------------------------------- combat
    def _perform_attack(self, att, dfn):
        if att.charge <= FULL_STRENGTH:
            return
        if not _ray_hits(att, dfn):
            return
        scale = max(0.0, min(1.0, att.charge))
        dmg = ATTACK_DAMAGE * (0.2 + 0.8 * scale * scale)
        crit = (not att.on_ground) and att.vy < 0.0 and not att.sprinting
        if crit:
            dmg *= 1.5
        taken = dmg * PROTECTION_MULT
        if taken > 0.0:
            dfn.health -= taken
            self._knockback(att, dfn)
        att.charge = 0.0
        att.sprinting = False   # sprint-reset / w-tap on a landed hit

    def _knockback(self, att, dfn):
        dx, dz = dfn.x - att.x, dfn.z - att.z
        d = math.hypot(dx, dz) or 1e-6
        nx, nz = dx / d, dz / d
        strength = BASE_KNOCKBACK + (SPRINT_KNOCKBACK if att.sprinting else 0.0)
        dfn.vx = dfn.vx / 2.0 + nx * strength
        dfn.vz = dfn.vz / 2.0 + nz * strength
        if dfn.on_ground:
            dfn.vy = min(0.4, dfn.vy / 2.0 + strength)
            dfn.on_ground = False
        dfn.was_hit = True

    def _recharge(self, b):
        b.charge = min(1.0, b.charge + CHARGE_PER_TICK)

    # ---------------------------------------------------- scripted bot2
    def _opponent_ai(self):
        b1, b2 = self.bot1, self.bot2
        dx, dy, dz = b1.x - b2.x, b1.y - b2.y, b1.z - b2.z
        b2.yaw = math.degrees(math.atan2(dz, dx))
        b2.pitch = -math.degrees(math.atan2(dy, math.hypot(dx, dz)))
        yaw = math.radians(b2.yaw)
        walk = WALK_IMPULSE
        if self._stap_ticks > 0:
            self._stap_ticks -= 1
            b2.sprinting = False
            b2.vx -= math.cos(yaw) * walk
            b2.vz -= math.sin(yaw) * walk
        else:
            b2.sprinting = True
            b2.vx += math.cos(yaw) * walk * SPRINT_MULTIPLIER
            b2.vz += math.sin(yaw) * walk * SPRINT_MULTIPLIER
        before = b2.charge
        self._perform_attack(b2, b1)
        if b2.charge < before:
            sampled = S_TAP_MEAN + self.rng.normal() * S_TAP_STD
            self._stap_ticks = int(round(max(1.0, min(float(S_TAP_MAX), sampled))))

    # ------------------------------------------------------- physics
    def _physics(self, b):
        b.x += b.vx
        b.y += b.vy
        b.z += b.vz
        b.vx *= H_FRICTION
        b.vz *= H_FRICTION
        if b.y <= 0.0:
            b.y = 0.0
            b.vy = 0.0
            b.on_ground = True
        else:
            b.vy = (b.vy - GRAVITY) * V_DRAG
            b.on_ground = False

    # ------------------------------------------------------- observation
    def _dist(self):
        return math.hypot(self.bot2.x - self.bot1.x, self.bot2.z - self.bot1.z)

    def _aim_error(self):
        """Angle (radians) between bot1's look vector and the direction to bot2's eye."""
        b1, b2 = self.bot1, self.bot2
        tx, ty, tz = b2.x - b1.x, (b2.y + EYE_HEIGHT) - (b1.y + EYE_HEIGHT), b2.z - b1.z
        tl = math.sqrt(tx * tx + ty * ty + tz * tz) or 1e-6
        lx, ly, lz = _look_vec(b1.yaw, b1.pitch)
        dot = (tx * lx + ty * ly + tz * lz) / tl
        return math.acos(max(-1.0, min(1.0, dot)))

    def _observe(self):
        b1, b2 = self.bot1, self.bot2
        dx, dz = b2.x - b1.x, b2.z - b1.z
        dy = b2.y - b1.y
        dist = math.hypot(dx, dz)
        # relative yaw to target
        bearing = math.degrees(math.atan2(dz, dx))
        rel_yaw = math.radians(_wrap_deg(bearing - b1.yaw))
        # relative pitch error (desired pitch onto target eye vs current pitch)
        desired_pitch = -math.degrees(math.atan2((b2.y + EYE_HEIGHT) - (b1.y + EYE_HEIGHT), max(dist, 1e-6)))
        rel_pitch = math.radians(_wrap_deg(desired_pitch - b1.pitch))
        # Facing-relative velocity (forward / strafe components) so it transfers cleanly to the
        # mod regardless of the sim-vs-MC yaw convention.
        syaw = math.radians(b1.yaw)
        fwd_v = b1.vx * math.cos(syaw) + b1.vz * math.sin(syaw)
        strafe_v = -b1.vx * math.sin(syaw) + b1.vz * math.cos(syaw)
        obs = np.array([
            min(dist / 10.0, 2.0),
            dy,
            math.sin(rel_yaw), math.cos(rel_yaw),
            math.sin(rel_pitch), math.cos(rel_pitch),
            fwd_v, strafe_v,
            b1.charge,
            b1.health / MAX_HEALTH,
            b2.health / MAX_HEALTH,
            1.0 if b1.on_ground else 0.0,
            1.0 if _ray_hits(b1, b2) else 0.0,
        ], dtype=np.float32)
        return obs

    # ------------------------------------------------------- reward
    def _reward(self, action, dmg_dealt, dmg_taken, could_hit):
        r = 0.0
        r += dmg_dealt
        r -= 0.75 * dmg_taken

        dist = self._dist()
        r += 0.02 * (self._prev_dist - dist)          # reward closing distance
        self._prev_dist = dist

        aim = self._aim_error()
        r += 0.1 * (self._prev_aim - aim)             # reward reducing aim error (drives 360+pitch)
        self._prev_aim = aim

        if action == ATTACK and not could_hit:
            r -= 0.1                                   # punish swinging when it can't connect

        if self.bot2.health <= 0:
            r += 20.0
        if self.bot1.health <= 0:
            r -= 20.0
        return r
